File: mobile/lib/domain/usecases/classify_device_usecase.dart
import '../entities/device.dart';
import '../repositories/device_repository.dart';

class ClassifyDeviceUseCase {
  final IDeviceRepository _deviceRepository;

  ClassifyDeviceUseCase(this._deviceRepository);

  Future<void> call(String macAddress, bool isKnown) async {
    final device = await _deviceRepository.getDeviceByMac(macAddress);
    if (device != null) {
      await _deviceRepository.updateDeviceStatus(macAddress, isKnown);
    } else {
      throw Exception('Device not found');
    }
  }
}

File: backend/src/domain/usecases/classify-device.usecase.ts
import { IDeviceRepository } from '../repositories/device.repository';
import { Device } from '../entities/device.entity';
import { DeviceCategory } from '../entities/network-device.entity';

export class ClassifyDeviceUseCase {
  constructor(private readonly deviceRepository: IDeviceRepository) {}

  async execute(macAddress: string): Promise<Device> {
    const device = await this.deviceRepository.findByMacAddress(macAddress);
    if (!device) {
      throw new Error('Device not found');
    }

    let category = DeviceCategory.UNKNOWN;
    if (device.vendor) {
      const vendorLower = device.vendor.toLowerCase();
      if (vendorLower.includes('apple') || vendorLower.includes('samsung') || vendorLower.includes('motorola')) {
        category = DeviceCategory.SMARTPHONE;
      } else if (vendorLower.includes('intel') || vendorLower.includes('dell') || vendorLower.includes('hp')) {
        category = DeviceCategory.COMPUTER;
      } else if (vendorLower.includes('espressif') || vendorLower.includes('tuya') || vendorLower.includes('philips')) {
        category = DeviceCategory.IOT;
      }
    }

    const updatedDevice = device.copy({ category });
    await this.deviceRepository.saveDevice(updatedDevice);
    return updatedDevice;
  }
}
